In [13]:
# Import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import fetch_openml  # Use fetch_openml to load the MNIST dataset
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score


In [14]:

# Load the MNIST dataset
mnist = fetch_openml('mnist_784', version=1)  # Use fetch_openml to load the MNIST dataset


C:\Users\match\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\datasets\_openml.py:1022: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [15]:

# Convert string labels to integers
mnist.target = mnist.target.astype(int)


In [16]:

# Split the data into features (X) and labels (y)
X = mnist.data
y = mnist.target


In [17]:

# Split the data into training and test sets
# Add a comment explaining the purpose of the train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Purpose of the train and test sets: Training set is used to train the model, 
# and the test set is used to evaluate the model's performance on unseen data.


In [18]:

# Create a Random Forest Classifier
# Pick one parameter to tune, and explain why you chose this parameter
# Chosen parameter: n_estimators (number of trees in the forest)
# Explanation: Increasing the number of trees can improve the model's accuracy, but it also increases the computational cost.


In [19]:

# Choose a value for the parameter for testing on the test data and explain why
n_estimators_value = 100
# Chosen value: 100, a common default value that generally provides a good balance between accuracy and computational efficiency.


In [20]:

# Initialize the Random Forest Classifier with the chosen parameter
rf_classifier = RandomForestClassifier(n_estimators=n_estimators_value, random_state=42)


In [21]:

# Train the model on the training data
rf_classifier.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

In [22]:

# Make predictions on the test data
y_pred = rf_classifier.predict(X_test)


In [23]:

# Print the confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)


Confusion Matrix:
[[1325    0    4    0    1    1    3    1    6    2]
 [   0 1573    6    8    2    0    0    6    3    2]
 [   4    5 1334    4    5    1    9    8    8    2]
 [   1    0   23 1366    0    9    0   14   12    8]
 [   4    1    3    0 1257    0    2    3    3   22]
 [   1    3    3   17    4 1226    8    1    8    2]
 [   4    1    0    0    6   10 1372    0    3    0]
 [   3    5   15    0    7    1    0 1455    2   15]
 [   1    6    9   15    6   11    5    7 1290    7]
 [   4    6    5   16   20    5    1   10    9 1344]]


In [24]:
# Report accuracy, precision, recall, and f1-score
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro')
recall = recall_score(y_test, y_pred, average='macro')
f1 = f1_score(y_test, y_pred, average='macro')

print("\nMetrics:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1 Score: {f1:.4f}")



Metrics:
Accuracy: 0.9673
Precision: 0.9671
Recall: 0.9671
F1 Score: 0.9671


Accuracy: 0.9673

This represents the overall correctness of the model across all classes.
In this case, 96.73% of the predictions made by the model on the test set were correct.
Precision: 0.9671

Precision is the ratio of correctly predicted positive observations to the total predicted positives.
A high precision value indicates that when the model predicts a certain class, it is usually correct.
In this case, on average, when the model predicts a positive instance, it is correct 96.71% of the time.
Recall: 0.9671

Recall (Sensitivity or True Positive Rate) is the ratio of correctly predicted positive observations to the all observations in the actual class.
A high recall value indicates that the model is good at capturing all instances of a particular class.
In this case, on average, the model captures 96.71% of the true positive instances for each class.
F1 Score: 0.9671

The F1 Score is the weighted average of precision and recall. It takes both false positives and false negatives into account.
A high F1 score indicates a good balance between precision and recall.
In this case, the F1 score is 96.71%, reflecting a good balance between precision and recall.
Overall, the model seems to perform well with high accuracy, precision, recall, and F1 score.